### ISSUE TITLE
Allow selecting "clobber" behavior for folders

### ISSUE THREAD

**Issue by GeorgeLyon** [sifive/wake#1152]
Right now, if my wake job outputs files, wake smashes those files into the existing filesystem. This could lead to some confusing results: for example, say I have a unit test running with `wake` and outputting artifacts to `build/test/scalatest/GeorgeTest` such that the first invocation of this test yields:
```
build/test/scalatest/GeorgeTest/simulation-log.txt
```
Now imagine this behavior is change to output the log as a `csv` instead. Either I could have made this change, or someone else might have and I happened to get it by rebasing my branch on top of master. Now if I invoke `wake` the test creates the following output:
```
build/test/scalatest/GeorgeTest/simulation-log.csv
```
Which wake happily smashes into my existing filesystem, yielding;
```
build/test/scalatest/GeorgeTest/simulation-log.txt (from the first invocation)
build/test/scalatest/GeorgeTest/simulation-log.csv (from the second invocation)
```
Now my test artifacts have inconsistent data, and if I was doing something like `cat build/test/scalatest/GeorgeTest/simulation-log.txt`, I may not even notice that this file is now out-of-date which could be very confusing during debugging.

The problem here is, I have no way to, when making the change to `simulation-log.csv`, tell `wake` to remove `simulation-log.txt`. It is up to the developer to magically know to delete that directory. It would be great if there was functionality to treat folders as files, and have `build/test/scalatest/GeorgeTest` be _completely replaced_ ("clobbered") when the second invocation runs. I mean, wouldn't it be pretty weird if two runs outputting `simulation-log.txt` _appended_ data to whatever was in `simulation-log.txt` on the filesystem?

Since most wake outputs are auto-detected, my napkin-sketch of a solution is this: search for a file called `.wakeattributes` in output folders, if that file has the line "clobber=true", then if that folder exists on the filesystem, wake will _replace the entire contents_ of the folder when writing the output files to the filesystem. This way, I can have my testing code add the `.wakeattributes` file to my test directory (whose name, by the way, I may not know prior to invoking the wake job) ensuring that subsequent changes to the layout of this folder don't get smashed with potentially-outdated data on the filesystem.

**Comment by ag-eitilt** [comment]
As another potential solution, if we don't want to do the `.wakeattributes` (though that might tie in well with the `.wakeroot` discussion), it might be possible to implement as a `mkdirClean` or something which does the emptying before returning the `Path`.  Either way we're going to have to figure out what to do with source files, directories where some other job has written/will write an output during the same `wake` invocation, etc.

**Comment by V-FEXrt** [comment]
Spitballing another idea. It may be possible to have wake automatically delete any output files of a `stale` job right before deleting it from the database.

I'm not a huge fan of deleting files from disk automatically but from wakes perspective, deleting them from the database is the same as deleting them from disk.

Though I'm sure some users indirectly depend on us not deleting stale outputs

**Comment by richardxia** [comment]
That sounds reasonable to me. I agree that it feels dangerous to automatically delete files and that some users might be indirectly depending on wake not deleting stale outputs, but I think it would even be reasonable to have to opt into this behavior for individual jobs.

In the scenario that I explained in the SiFive-internal ticket, we're talking about Python virtual environments, where stale files in the virtual environment may cause really strange behavior because you'd have both old and new files for a package getting mixed together. I think it would be totally reasonable for the build rule that produced the virtual environment to declare that it should opt into this "delete files from stale job" behavior, which I don't think any user would depend on.

**Comment by V-FEXrt** [comment]
Yeah, opting-in would be a good solution here. That way the job can make stronger guarantees. Doing that requires a bit more plumbing (`Plan` all the way down into the db), but it would be a good improvement to have.

I think I'm starting to rack up a few changes that would require incrementing the db schema. Probably good to throw this one on that stack.

**Comment by ag-eitilt** [comment]
I don't think this will affect things in any significant way, but just as something to keep in mind while implementing it: the stale job might be stale because an *entirely different* job has run instead and just resulted in the same files, rather than the same job just being passed different inputs.  One situation where I (ab)use `claim` to skip processing on subsequent runs winds up marking several source jobs as being stale, so it's not even guaranteed to be one-to-one.  So far as I can imagine, that would only wind up making it a bit more challenging to pick what Plan's opt-in should be used, especially if the old and new Job outputs only partially overlap.

**Comment by ag-eitilt** [comment]
Reopening this because we're pointing to it as the tracking issue for this request, which keeps coming up.  Recent discussion makes us think that this *might* be able to be done safely if we only delete files that wake itself created in a previous run, though race conditions are still tricky.
